### PR TITLE
Initial changes for KNearestNeighbor Classifier/Regressor and PCA

### PR DESCRIPTION
**Pull request by prabhat00155**
Add KNNClassifier, KNNRegressor, and PCA conversions from scikit-learn.

### CODE REVIEW

File: onnxmltools/convert/sklearn/shape_calculators/PCA.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# -------------------------------------------------------------------------
```
Comment (wschin) [inline onnxmltools/convert/sklearn/shape_calculators/PCA.py]: This file is not necessary. Just copy `register_shape_calculator('SklearnPCA', calculate_sklearn_pca_output_shapes)` to this [place](https://github.com/onnx/onnxmltools/blob/0cd78a742837aa429c57d3a54f269a2421770cca/onnxmltools/convert/sklearn/shape_calculators/SVD.py#L31) because PCA and SVD are conceptually the same.

File: onnxmltools/convert/sklearn/shape_calculators/KNeighborsClassifier.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# -------------------------------------------------------------------------
```
Comment (wschin) [inline onnxmltools/convert/sklearn/shape_calculators/KNeighborsClassifier.py]: It looks like this file is not necessary as well. Could you reuse [this shape calculator](https://github.com/onnx/onnxmltools/blob/0cd78a742837aa429c57d3a54f269a2421770cca/onnxmltools/convert/sklearn/shape_calculators/LinearClassifier.py#L69) by moving `



 
register_shape_calculator('SklearnKNeighborsClassifier', calculate_sklearn_k_neighbors_classifier_output_shapes)`?

File: onnxmltools/convert/sklearn/operator_converters/PCA.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# -------------------------------------------------------------------------
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/PCA.py]: Could you merge this into the converter of scikit-learn SVD? They are almost identical.

File: onnxmltools/convert/sklearn/operator_converters/KNeighborsClassifier.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
+# -------------------------------------------------------------------------
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/KNeighborsClassifier.py]: Is it possible to merge KNNClassifier and KNNRegressor converters? Their computation graphs are quite similar. I guess we can add an if-else for creating the nodes after selecting top-k examples, so that the code for creating common structures in KNNC and KNNR can be shared.

File: onnxmltools/convert/sklearn/shape_calculators/KNeighborsClassifier.py (hunk 2 above)
Comment (prabhat00155) [reply onnxmltools/convert/sklearn/shape_calculators/KNeighborsClassifier.py]: Done

File: onnxmltools/convert/sklearn/operator_converters/KNeighborsClassifier.py (hunk 4 above)
Comment (prabhat00155) [reply onnxmltools/convert/sklearn/operator_converters/KNeighborsClassifier.py]: Done.

File: onnxmltools/convert/sklearn/operator_converters/KNN.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+                                  classes.shape, classes)
+        container.add_initializer(training_labels_name, class_type,
+                                  training_labels.shape, training_labels)
+
+        container.add_node('ArrayFeatureExtractor', [training_labels_name, topk_indices_name],
+                           topk_labels_name, name='ArrayFeatureExtractor1', op_domain='ai.onnx.ml')
+        for i in range(len(classes)):
+            container.add_node('Equal', [labels_name[i], topk_labels_name],
+                                output_label_name[i], op_version=7)
+            container.add_node('Cast', output_label_name[i],
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/KNN.py]: Why Int32? Number of examples should be `Int64` by default.

File: onnxmltools/convert/sklearn/operator_converters/LeastSquares.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+from ....proto import onnx_proto
+from ...common._registration import register_converter
+
+
+def convert_sklearn_least_squares(scope, operator, container):
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/LeastSquares.py]: Looks like you can reuse the converter for [linear regression](https://github.com/onnx/onnxmltools/blob/master/onnxmltools/convert/sklearn/operator_converters/LinearRegressor.py).

File: onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    container.add_initializer(shape_name, onnx_proto.TensorProto.INT64,
+                              [len(shape)], shape)
+    container.add_initializer(output_shape_name, onnx_proto.TensorProto.INT64,
+                              [len(output_shape)], output_shape)
+    container.add_initializer(class_log_prior_name, onnx_proto.TensorProto.FLOAT,
+                              class_log_prior.shape, class_log_prior)
+    container.add_initializer(feature_log_prob_name, onnx_proto.TensorProto.FLOAT,
+                              feature_log_prob.shape, feature_log_prob.flatten())
+
+    container.add_node('Reshape', [class_log_prior_name, shape_name],
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py]: class_log_prior_name and shape_name are names of static variables. Can we just compute the result here instead of asking runtime to do the Reshape?

File: onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+from ....proto import onnx_proto
+from ...common._registration import register_converter
+
+
+def convert_sklearn_naive_bayes(scope, operator, container):
+    # Multinomial NB
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py]: Could you also draw the computational graph with all shape specified? I am not sure if you need those Reshape's below and would like to have more information about your conversion. You're using the shape calculator for linear classifier (which produces two outputs) the first equation below only shows one output.

File: onnxmltools/convert/sklearn/shape_calculators/LinearClassifier.py [hunk 9]
```diff
@@ -36,26 +36,28 @@ def calculate_sklearn_linear_classifier_output_shapes(operator):
         class_labels = np.concatenate(class_labels)
     if all(isinstance(i, (six.string_types, six.text_type)) for i in class_labels):
         operator.outputs[0].type = StringTensorType(shape=[N])
-        if len(class_labels) > 2 or operator.type != 'SklearnLinearSVC':
-            # For multi-class classifier, we produce a map for encoding the probabilities of all classes
-            if operator.targeted_onnx_version < StrictVersion('1.2'):
-                operator.outputs[1].type = DictionaryType(StringTensorType([1]), FloatTensorType([1]))
+        if operator.type != 'SklearnKNeighborsClassifier' and 'NB' not in operator.type:
```
Comment (wschin) [inline onnxmltools/convert/sklearn/shape_calculators/LinearClassifier.py]: Did you change the shape calculation rule described right below this function's header? This shape calculator requires that your classifier to have [N, C] input shape, so you need to accept [N, C] and reshape it into [C] where N is always 1 in your graph. Also, there are always two outputs, [one for label](http://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.BernoulliNB.html#sklearn.naive_bayes.BernoulliNB.predict) and the other [one for probability](http://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.BernoulliNB.html#sklearn.naive_bayes.BernoulliNB.predict_proba). Consequently, I don't think the shape calculation rules should be changed here.
Note that an important concept is that shape calculator determines the input and output shape of your graph, so the associated converter should make sure the produced sub-graph meets the shapes produced here.

File: tests/sklearn/test_KNNConverter.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
+import unittest
+from sklearn.neighbors import KNeighborsRegressor
+from sklearn.datasets import load_iris
+from onnxmltools import convert_sklearn
+from onnxmltools.convert.common.data_types import FloatTensorType
+
+
+class TestKNNConverter(unittest.TestCase):
+
+    def _fit_model_multiclass_classification(self, model):
```
Comment (wschin) [inline tests/sklearn/test_KNNConverter.py]: Maybe _fit_knn is a better name?

File: onnxmltools/convert/sklearn/operator_converters/KNN.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+                       distance_name, name=scope.get_unique_operator_name('Pow'))
+    container.add_node('ReduceSum', distance_name,
+                       reduced_sum_name, name=scope.get_unique_operator_name('ReduceSum'), axes=[1])
+    container.add_node('Reshape', [reduced_sum_name, length_name],
+                       reshaped_name, name=scope.get_unique_operator_name('Reshape'))
+    container.add_node('TopK', reshaped_name,
+                       [topk_values_name, topk_indices_name], name=scope.get_unique_operator_name('TopK'), k=knn.n_neighbors)
+
+    if operator.type == 'SklearnKNeighborsClassifier':
+        # To be implemented
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/KNN.py]: Please throw.

File: onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    # Equation: 
+    #   y = argmax (class_log_prior + X . feature_log_prob^T)
+    #
+    # Graph:
+    #
+    #   input[M, N] -> MATMUL <- feature_log_prob.T[N, C]
+    #                    |
+    #                    V
+    #        matmul_result[M, C] -> CAST <- onnx_proto.TensorProto.FLOAT
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py]: ><- onnx_proto.TensorProto.FLOAT [](start = 41, length = 31)

Is it an input? I thought it was an attribute so it could be ignored.

File: onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+    #                    cast_result[M, C] -> SUM <- class_log_prior[1, C]
+    #                                          |
+    #                                          V                                    (string labels)
+    #                             sum_result[M, C] -> ARGMAX -> argmax_output[M, 1] --------------------|
+    #                                               (int labels) |                                      |
+    #                                                            V                                      |
+    #      output_shape[1] -> RESHAPE <- cast2_result[M, 1] <- CAST <- onnx_proto.TensorProto.FLOAT     |
+    #                          |                                                                        |
+    #                          V                                                                        V
+    #                       reshaped_result[M,]             |------------------------------------- RESHAPE
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py]: No need to have Reshape. By setting keepdims=0 for ARGMAX, we may have output shape [M,]

File: onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+    #  feature_log_prob.T[N, C] -> EXP -> exp_result[N, C] 
+    #                                      |
+    #                                      V
+    #                         constant -> SUB -> sub_result[N, C] -> LOG -> neg_prob[N, C]
+    #                                                                        |
+    #                                                                        V
+    #  ----------------- sum_neg_prob[1, C] <------------------------ REDUCE_SUM
+    #  |                     |
+    #  |                     V
+    #  |    input[M, N] -> MATMUL -> inp_neg_prob_prod[M, C] -> CAST <- onnx_proto.TensorProto.FLOAT
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py]: > onnx_proto.TensorProto.FLOAT [](start = 71, length = 29)

It's not an input. Please remove it. Thanks.

File: onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+    #                                  array_feature_extractor_result[M, 1] ----------------------------
+    #                                               (int labels) |                                     | 
+    #                                                            V                                     |
+    #      output_shape[1] -> RESHAPE <- cast2_result[M, 1] <- CAST <- onnx_proto.TensorProto.FLOAT    | 
+    #                          |                                                                       |
+    #                          V                                                                       |
+    #                       reshaped_result[M,]             --------------------------------------RESHAPE
+    #                                   |                   |
+    #                                   V                   |
+    #  onnx_proto.TensorProto.INT64 -> CAST -> output[M,] <-|
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py]: > -> output[M,] <-| [](start = 43, length = 18)

I don't quite understand. Is it a double-assignment?

File: onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+
+    nb = operator.raw_operator
+    class_log_prior = nb.class_log_prior_.astype('float32').reshape((1, -1))
+    feature_log_prob = nb.feature_log_prob_.T.astype('float32')
+    classes = nb.classes_
+    output_shape = [-1,]
+
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py]: It looks like Flatten fits your need better: [https://aiinfra.visualstudio.com/Lotus/Lotus%20Team/_git/Lotus?path=%2Fonnxruntime%2Ftest%2Fpython%2Fonnxruntime_test_python.py&version=GBmaster&line=147&lineStyle=plain&lineEnd=148&lineStartColumn=1&lineEndColumn=1](https://aiinfra.visualstudio.com/Lotus/Lotus%20Team/_git/Lotus?path=%2Fonnxruntime%2Ftest%2Fpython%2Fonnxruntime_test_python.py&version=GBmaster&line=147&lineStyle=plain&lineEnd=148&lineStartColumn=1&lineEndColumn=1).

File: onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+        inp_neg_prob_prod_name = scope.get_unique_variable_name('inp_neg_prob_prod')
+        difference_matrix_name = scope.get_unique_variable_name('difference_matrix')
+
+        container.add_initializer(constant_name, onnx_proto.TensorProto.FLOAT,
+                                  [], [1.0])
+
+        container.add_node('Exp', feature_log_prob_name,
+                           exp_result_name, name=scope.get_unique_operator_name('Exp'), op_version=6)
+        container.add_node('Sub', [constant_name, exp_result_name],
+                           sub_result_name, name=scope.get_unique_operator_name('Sub1'))
```
Comment (wschin) [inline onnxmltools/convert/sklearn/operator_converters/NaiveBayes.py]: To support multiple-onnx versions, please use this function: [https://github.com/onnx/onnxmltools/blob/5aabc64d6ebe88e96a8f4f5949db404b10ae2218/onnxmltools/convert/common/_apply_operation.py#L261](https://github.com/onnx/onnxmltools/blob/5aabc64d6ebe88e96a8f4f5949db404b10ae2218/onnxmltools/convert/common/_apply_operation.py#L261)

File: tests/sklearn/test_LeastSquaresConverter.py [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+
+    def test_model_lasso_lars(self):
+        model = self._fit_model_least_squares(LassoLars(alpha=0.01))
+        model_onnx = convert_sklearn(model, 'lasso lars', [('input', FloatTensorType([1, 2]))])
+        self.assertIsNotNone(model_onnx)
+
+    def test_model_ridge(self):
+        model = self._fit_model_least_squares(Ridge())
+        model_onnx = convert_sklearn(model, 'ridge', [('input', FloatTensorType([1, 2]))])
+        self.assertIsNotNone(model_onnx)
```
Comment (wschin) [inline tests/sklearn/test_LeastSquaresConverter.py]: May we merge these two tests into this file: [https://github.com/onnx/onnxmltools/blob/master/tests/sklearn/test_GLMRegressorConverter.py](https://github.com/onnx/onnxmltools/blob/master/tests/sklearn/test_GLMRegressorConverter.py)? Thanks.
